File: 3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0: return 0
        left = win = 0
        chat_index = {} # 记录位置，这个最关键
        for right in range(len(s)):
            while s[right] in chat_index and chat_index[s[right]] >= left:
                left = chat_index[s[right]] + 1
            win = max(win, right-left+1)
            chat_index[s[right]] = right
        return win






# @lc code=end

File: 3.无重复字符的最长子串.java
/*
 * @lc app=leetcode.cn id=3 lang=java
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start

import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length()<=1)return s.length();
        HashMap<Character, Integer> map = new HashMap<>();
        int win = 0;
        int left = 0; 
        int n = s.length();
        for(int right=0;right<n;right++){
            char ch = s.charAt(right);
            if(map.containsKey(ch)&&map.get(ch)>=left){ //验证未被排除在窗口之外
                left = map.get(ch)+1;
            }else{
                win = Math.max(win, right-left+1);
            }
            map.put(ch, right); //一次只会重复一个，旧的已经被排除了
        }
        return win;
    }
}
// @lc code=end

